src/charm.py

```diff
@@ -303,17 +300,43 @@ def _current_mimir_config(self) -> dict:
             )
             return {}
 
-    @property
-    def _mimir_version(self) -> Optional[str]:
-        if not self._container.can_connect():
-            return None
+    def _set_alerts(self) -> None:
+        """(Re-)create alert rule files for all Mimir consumers.
 
-        version_output, _ = self._container.exec(["/bin/mimir", "-version"]).wait_output()
-        # Output looks like this:
-        # Mimir, version 2.4.0 (branch: HEAD, revision: 32137ee)
-        if result := re.search(r"[Vv]ersion:?\s*(\S+)", version_output):
-            return result.group(1)
-        return None
+        Returns: True if alerts rules have changed, otherwise False.
+        Raises: BlockedStatusError exception if PebbleError, ProtocolError, PathError exceptions
+            are raised by container.remove_path
+        """
+        try:
+            self._container.remove_path(RULES_DIR, recursive=True)
+        except PebbleError as e:
+            logger.error("Failed to remove alerts directory: %s", e)
+            raise BlockedStatusError("Failed to remove alerts directory; see debug logs")
+
+        try:
+            self._push_alert_rules(self.remote_write_provider.alerts())
+        except (ProtocolError, PathError) as e:
+            logger.error("Failed to push updated alert files: %s", e)
+            raise BlockedStatusError("Failed to push updated alert files; see debug logs")
+
+    def _push_alert_rules(self, alerts):
+        """Push alert rules from a rules file to the mimir container.
+
+        Args:
+            alerts: a dictionary of alert rule files.
+        """
+        # Without multitenancy, the default is `anonymous`, and the ruler checks under
+        # {RULES_DIR}/<tenant_id>
+        tenant_dir = f"{RULES_DIR}/anonymous"
+        # Need to `make_dir` even though we have `make_dirs=True` below
+        # https://github.com/canonical/operator/issues/898
+        self._container.make_dir(tenant_dir, make_parents=True)
+        for topology_identifier, rules_file in alerts.items():
+            filename = f"juju_{topology_identifier}.rules"
+            path = os.path.join(tenant_dir, filename)
+            rules = yaml.safe_dump(rules_file)
```

Comment:
rules_file is `str` then? I'd give the variable a different name then, and update the docstring accordingly. A type hint would also be nice.

Reply:
Agreed. This is just a re-shuffle. I'll follow back through for cleanup.